text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+        * Usually requires editing the trait anyway, so there's no "can constify impls without trait author opt in" silver bullet.
+    * New RTN-like per-method bounds: `T: Trait<some_fn(_): (const)>`.
+        * Unclear if soundly possible.
+        * Unclear if possible without incurring significant performance issues for all code (may need tracking new information for all functions out there).
+        * Still requires editing traits.
+        * Still want the `(const) Trait` sugar anyway.
+
+## Should we start out without `const Trait` bounds
+
+We do not need to immediately allow using methods on generic parameters of const fn, as a lot of const code is nongeneric.
+
+The following example could be made to work with just const traits and const trait impls.
+
+```rust
+struct MyStruct(i32);
+
+impl PartialEq for MyStruct {
+    const fn eq(&self, other: &MyStruct) -> bool {
+        self.0 == other.0
+    }
+}
+
+const fn foo() {
+    let a = MyStruct(1);
+    let b = MyStruct(2);
+    if a == b {}
+}
+```
+
+Things like `Option::map` or `PartialEq` for arrays/tuples could not be made const without const trait bounds,
+as they need to actually call the generic `FnOnce` argument or nested `PartialEq` impls.
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Migrate to `(const) fn`
+
+`const fn` and `const` items have slightly different meanings for `const`:
+
+`const fn` can also be called at runtime just fine, while the others are always const
+contexts and need to be evaluated by the const evaluator.
+
+Additionally `const Trait` bounds have a third meaning (the same as `const Trait` in `impl const Trait for Type`):
+
+They can be invoked at compile time, but also in `const fn`.
+
+While all these meanings are subtly different, making their differences more obvious will not make them easier to understand.
+All that changing to `(const) fn` would achieve is that folk will add the sigil when told by the compiler, and complain about
+having to type a sigil, when there is no meaning for `const fn` without a sigil.
+
+While I see the allure from a language nerd perspective to give every meaning its own syntax, I believe it is much more practical to
+just call all of these `const` and only separate the `(const) Trait` bounds from `const Trait` bounds.
+
+## `const fn()` pointers
+
+Just like `const fn foo(x: impl (const) Trait) { x.method() }` and `const fn foo(x: &dyn (const) Trait) { x.method() }` we want to allow
+`const fn foo(f: (const) fn()) { f() }`.
+
+These require changing the type system, making the constness of a function pointer part of the type.
+This in turn implies that a `const fn()` function pointer, a `(const) fn()` function pointer and a `fn()` function pointer could have
+different `TypeId`s, which is something that requires more design and consideration to clarify whether supporting downcasting with `Any`
+or just supporting `TypeId` equality checks detecting constness is desirable.
```

Comment:
It gets a lot more complicated than that: surely we also want people to be able to newtype such a function pointer type, so all of these are now a thing:
```rust
struct S1(fn());
struct S2(const fn());
struct S3((const) fn());
```
From here it is actually not far to `(const) Trait` bounds on types (which the RFC still claims make no sense, which is IMO an incorrect claim).

Reply:
With types like `struct S3((const) fn());`, clearly what we call a "type" is now more like a "function from constness to types". And thus the `where` clauses in a type can refer to this ambient constness via `(const) Trait`, similar to `where` clauses in a `const fn`.

Reply:
Note that just because there is motivation for `const`-parameterized user defined types, that does not necessarily imply that we should also allow `(const) Trait` bounds on said types.

Reply:
Note that just because there is motivation for `const`-parameterized user-defined types, that does not necessarily imply that we should also allow `(const) Trait` bounds on said types.

Reply:
Correct, and I did not say that it would imply as much. But "such bounds make conceptually no sense" is factually incorrect. If we dismiss such bounds we should do it properly. The complexity of "a type's meaning depends on the surrounding context" will inevitably become part of the language once we have `(const) fn` types or `dyn (const) Trait` types, so at that point I don't see a good reason to prevent the user from also using this in where clauses.